app/src/foundation/ContentScript/ContentScript.ts

```diff
@@ -1,179 +1,76 @@
-import {createEndpoint, retain} from '@remote-ui/rpc';
-import type {Endpoint} from '@remote-ui/rpc';
-import {
-  ContentScriptApiForWebpage,
-  ContentScriptApiForDevtools,
-  DevtoolsApiForContentScript,
-  fromWebpage,
-  WebpageApi,
-} from '@aside/core';
 import {Runtime} from 'webextension-polyfill';
-import {type RemoteChannel} from '@remote-ui/core';
+import {effect} from '@preact/signals-core';
 
-import {createUnsafeEncoder} from '../Remote';
-import {fromPort} from '../Remote/rpc';
-
-import {WEBPAGE_INITIATED_CONNECTION} from './messages';
-
-interface Current {
-  webpage?: Endpoint<WebpageApi>;
-  devtools?: Endpoint<DevtoolsApiForContentScript>;
-  port?: Runtime.Port;
-}
+import {createReducer} from './reducer';
 
 export function contentScript() {
-  const current: Current = {};
+  const [stateSignal, dispatch] = createReducer({
```

Comment:
Having the app/content-script connection state into a signal would make it very simply to create a subscribable and turn this into a declarative api. Whenever the webpage/content-script/devtools would connect to each other they would receive a subscribable with the connection state and would be able to react to it without any imperative messages.